File: python-scripts/edit_blueprint_graph.py
"""
Apply a batch of Blueprint graph edits in a single editor session, then compile + save.

Driven by the `edit-blueprint-graph` MCP tool. Calls UClaudeBPGraphLibrary (shipped by
the ClaudeUnrealBridge editor plugin; the sandbox spike plugin exposes the same class),
which wraps the standard UE editor graph APIs (FGraphNodeCreator, schema TryCreateConnection,
UEdGraph::RemoveNode, FKismetEditorUtilities::CompileBlueprint).

Args (via builtins._claude_args):
  asset_path   : /Game/... path to the Blueprint
  graph_name   : graph to edit; "" = the event graph (default)
  operations   : JSON array of ops (see below)
  compile      : "true"/"false" (default true) — compile after applying ops
  output_file  : where to write the JSON result

Operation kinds (each a JSON object with "op"):
  {"op":"addFunctionNode","id":"h1","functionOwner":"SystemLibrary","functionName":"PrintString","x":0,"y":0}
  {"op":"addCustomEvent","id":"evt","eventName":"MyEvent","x":-300,"y":0}
  {"op":"connect","from":"evt","fromPin":"then","to":"h1","toPin":"execute"}
  {"op":"setPinDefault","node":"h1","pin":"InString","value":"Hello"}
  {"op":"deleteNode","node":"<localId-or-existing-GUID>"}

Node references ("from"/"to"/"node") resolve against local ids assigned by earlier add ops;
anything not a known local id is treated as an existing node GUID (from read-blueprint).
"""
import json
import traceback
import builtins

import unreal


def get_args():
    return getattr(builtins, "_claude_args", {})


def main():
    args = get_args()
    asset_path = args.get("asset_path", "")
    graph_name = args.get("graph_name", "")
    output_file = args.get("output_file")
    do_compile = str(args.get("compile", "true")).lower() != "false"
    auto_layout = str(args.get("auto_layout", "false")).lower() == "true"
    try:
        ops = json.loads(args.get("operations", "[]"))
    except Exception as e:
        ops = []

    result = {
        "success": False,
        "asset_path": asset_path,
        "graph": graph_name,
        "operations": [],
        "handles": {},
        "compiled": None,
        "saved": None,
        "nodes": [],
    }

    GraphLib = getattr(unreal, "ClaudeBPGraphLibrary", None)
    if GraphLib is None:
        result["error"] = (
            "UClaudeBPGraphLibrary is not available. Enable the ClaudeUnrealBridge "
            "plugin in this project and build its editor target (the ClaudeUnrealBridgeEditor "
            "module) once, then retry."
        )
        _write(output_file, result)
        return

    bp = unreal.EditorAssetLibrary.load_asset(asset_path)
    if bp is None or not isinstance(bp, unreal.Blueprint):
        result["error"] = "Asset is not a Blueprint (or not found): %s" % asset_path
        _write(output_file, result)
        return

    handles = {}

    def resolve_node(ref):
        return handles.get(ref, ref)

    def resolve_class(name):
        if not name:
            return None
        # Full object path, e.g. "/Script/Engine.KismetSystemLibrary" or a BP class path.
        if name.startswith("/"):
            obj = unreal.load_object(None, name)
            return obj
        # Python-exposed class name, e.g. "SystemLibrary" (== UKismetSystemLibrary),
        # "GameplayStatics", "MathLibrary", or a user UCLASS.
        obj = getattr(unreal, name, None)
        if obj is not None and hasattr(obj, "static_class"):
            return obj.static_class()
        if hasattr(unreal, "find_class"):
            try:
                return unreal.find_class(name)
            except Exception:
                return None
        return None

    for i, op in enumerate(ops):
        rec = {"index": i, "op": op.get("op", "?")}
        try:
            kind = op.get("op")
            if kind == "addFunctionNode":
                owner = resolve_class(op.get("functionOwner", ""))
                if owner is None:
                    raise ValueError("could not resolve functionOwner '%s'" % op.get("functionOwner"))
                guid = GraphLib.add_function_call_node(
                    bp, graph_name, owner, op["functionName"],
                    int(op.get("x", 0)), int(op.get("y", 0)))
                rec["guid"] = guid
                rec["ok"] = bool(guid)
                if op.get("id") and guid:
                    handles[op["id"]] = guid
            elif kind == "addCustomEvent":
                guid = GraphLib.add_custom_event_node(
                    bp, graph_name, op["eventName"],
                    int(op.get("x", 0)), int(op.get("y", 0)))
                rec["guid"] = guid
                rec["ok"] = bool(guid)
                if op.get("id") and guid:
                    handles[op["id"]] = guid
            elif kind == "connect":
                rec["ok"] = bool(GraphLib.connect_pins(
                    bp, graph_name, resolve_node(op["from"]), op["fromPin"],
                    resolve_node(op["to"]), op["toPin"]))
            elif kind == "setPinDefault":
                rec["ok"] = bool(GraphLib.set_pin_default(
                    bp, graph_name, resolve_node(op["node"]), op["pin"], op["value"]))
            elif kind == "deleteNode":
                rec["ok"] = bool(GraphLib.delete_node(
                    bp, graph_name, resolve_node(op["node"])))
            elif kind == "addMemberVariable":
                rec["ok"] = bool(GraphLib.add_member_variable(
                    bp, op["name"], op.get("varType", "int"), str(op.get("default", ""))))
            elif kind == "addVariableGet":
                guid = GraphLib.add_variable_get_node(
                    bp, graph_name, op["variable"],
                    int(op.get("x", 0)), int(op.get("y", 0)))
                rec["guid"] = guid
                rec["ok"] = bool(guid)
                if op.get("id") and guid:
                    handles[op["id"]] = guid
            elif kind == "addVariableSet":
                guid = GraphLib.add_variable_set_node(
                    bp, graph_name, op["variable"],
                    int(op.get("x", 0)), int(op.get("y", 0)))
                rec["guid"] = guid
                rec["ok"] = bool(guid)
                if op.get("id") and guid:
                    handles[op["id"]] = guid
            elif kind == "addBranch":
                guid = GraphLib.add_branch_node(
                    bp, graph_name, int(op.get("x", 0)), int(op.get("y", 0)))
                rec["guid"] = guid
                rec["ok"] = bool(guid)
                if op.get("id") and guid:
                    handles[op["id"]] = guid
            elif kind == "addSequence":
                guid = GraphLib.add_sequence_node(
                    bp, graph_name, int(op.get("numOutputs", 2)),
                    int(op.get("x", 0)), int(op.get("y", 0)))
                rec["guid"] = guid
                rec["ok"] = bool(guid)
                if op.get("id") and guid:
                    handles[op["id"]] = guid
            elif kind == "addMacro":
                guid = GraphLib.add_macro_instance_node(
                    bp, graph_name, op["macro"],
                    int(op.get("x", 0)), int(op.get("y", 0)))
                rec["guid"] = guid
                rec["ok"] = bool(guid)
                if op.get("id") and guid:
                    handles[op["id"]] = guid
            elif kind == "addCast":
                owner = resolve_class(op.get("targetClass", ""))
                if owner is None:
                    raise ValueError("could not resolve targetClass '%s'" % op.get("targetClass"))
                guid = GraphLib.add_cast_node(
                    bp, graph_name, owner, bool(op.get("pure", False)),
                    int(op.get("x", 0)), int(op.get("y", 0)))
                rec["guid"] = guid
                rec["ok"] = bool(guid)
                if op.get("id") and guid:
                    handles[op["id"]] = guid
            elif kind == "addSelfFunctionNode":
                guid = GraphLib.add_self_function_call_node(
                    bp, graph_name, op["functionName"],
                    int(op.get("x", 0)), int(op.get("y", 0)))
                rec["guid"] = guid
                rec["ok"] = bool(guid)
                if op.get("id") and guid:
                    handles[op["id"]] = guid
            elif kind == "breakPinLink":
                rec["ok"] = bool(GraphLib.break_pin_link(
                    bp, graph_name, resolve_node(op["from"]), op["fromPin"],
                    resolve_node(op["to"]), op["toPin"]))
            elif kind == "retargetNode":
                node = resolve_node(op["node"])
                if "functionName" in op:
                    owner = resolve_class(op.get("functionOwner", ""))
                    if owner is None:
                        raise ValueError("retargetNode: could not resolve functionOwner '%s'" % op.get("functionOwner"))
                    rec["ok"] = bool(GraphLib.retarget_function_node(bp, graph_name, node, owner, op["functionName"]))
                elif "targetClass" in op:
                    owner = resolve_class(op["targetClass"])
                    if owner is None:
                        raise ValueError("retargetNode: could not resolve targetClass '%s'" % op.get("targetClass"))
                    rec["ok"] = bool(GraphLib.retarget_cast_node(bp, graph_name, node, owner))
                elif "variable" in op:
                    rec["ok"] = bool(GraphLib.retarget_variable_node(bp, graph_name, node, op["variable"]))
                elif "eventName" in op:
                    rec["ok"] = bool(GraphLib.rename_custom_event_node(bp, graph_name, node, op["eventName"]))
                else:
                    rec["ok"] = False
                    rec["error"] = "retargetNode needs one of: functionName(+functionOwner), targetClass, variable, eventName"
            elif kind == "moveNode":
                rec["ok"] = bool(GraphLib.set_node_position(
                    bp, graph_name, resolve_node(op["node"]),
                    int(op.get("x", 0)), int(op.get("y", 0))))
            else:
                rec["ok"] = False
                rec["error"] = "unknown op '%s'" % kind
        except Exception as e:
            rec["ok"] = False
            rec["error"] = "".join(traceback.format_exception_only(type(e), e)).strip()
        result["operations"].append(rec)

    if auto_layout:
        try:
            result["auto_layout"] = bool(GraphLib.auto_layout_graph(bp, graph_name))
        except Exception as e:
            result["auto_layout"] = False
            result["auto_layout_error"] = str(e)

    if do_compile:
        try:
            result["compiled"] = bool(GraphLib.compile_blueprint_asset(bp))
        except Exception as e:
            result["compiled"] = False
            result["compile_error"] = str(e)

    try:
        result["saved"] = bool(unreal.EditorAssetLibrary.save_loaded_asset(bp))
    except Exception as e:
        result["saved"] = False
        result["save_error"] = str(e)

    try:
        result["nodes"] = [str(s) for s in GraphLib.list_node_guids(bp, graph_name)]
    except Exception as e:
        result["nodes_error"] = str(e)

    result["handles"] = handles
    ops_ok = all(o.get("ok") for o in result["operations"]) if result["operations"] else True
    compile_ok = (result["compiled"] is not False) if do_compile else True
    result["success"] = bool(ops_ok and compile_ok)

    _write(output_file, result)


def _write(output_file, result):
    if output_file:
        with open(output_file, "w") as f:
            json.dump(result, f, indent=2, default=str)
    unreal.log("EDIT_BP_GRAPH_DONE success=%s" % result.get("success"))


main()

File: mcp-server/src/tools/edit-blueprint-graph.ts
import { z } from "zod";
import { join } from "node:path";
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { detectProject, normalizeToAssetPath } from "../ue-bridge/project-detector.js";
import { runPythonInUE } from "../ue-bridge/python-runner.js";
import { progressFromExtra } from "../mcp/progress.js";
import { graphEditStructuredShape } from "../mcp/output-schemas.js";

const pluginRoot = process.env.PLUGIN_ROOT || process.cwd();

interface OpResult {
  index: number;
  op: string;
  ok?: boolean;
  guid?: string;
  error?: string;
}
interface GraphEditResult {
  success: boolean;
  asset_path: string;
  graph: string;
  operations: OpResult[];
  handles: Record<string, string>;
  compiled?: boolean | null;
  saved?: boolean | null;
  auto_layout?: boolean | null;
  nodes?: string[];
  connections?: string[];
  error?: string;
  compile_error?: string;
  save_error?: string;
  auto_layout_error?: string;
}

/** Parse a "fromGuid|fromPin|toGuid|toPin|kind" edge string into an object. */
function parseConnection(s: string): {
  from: string;
  fromPin: string;
  to: string;
  toPin: string;
  kind: string;
} | null {
  const p = String(s).split("|");
  if (p.length < 5) return null;
  return { from: p[0], fromPin: p[1], to: p[2], toPin: p[3], kind: p[4] };
}

// Discriminated union of graph operations. Node references ("from"/"to"/"node")
// are either a local `id` assigned by an earlier add op in THIS batch, or an
// existing node GUID (from read-blueprint). Exec pins are named `then` (output)
// and `execute` (input); a function's data input pins use the UE parameter name.
const opSchema = z.discriminatedUnion("op", [
  z.object({
    op: z.literal("addFunctionNode"),
    id: z.string().optional().describe("Local handle for referencing this new node in later ops."),
    functionOwner: z
      .string()
      .describe("Class that owns the function: a UE Python class name (e.g. \"SystemLibrary\" = UKismetSystemLibrary, \"GameplayStatics\", \"MathLibrary\"), or a full class path (\"/Script/Engine.KismetSystemLibrary\", \"/Game/.../BP_Foo.BP_Foo_C\")."),
    functionName: z.string().describe("Function name, e.g. \"PrintString\"."),
    x: z.number().optional().describe("Node X position (default 0)."),
    y: z.number().optional().describe("Node Y position (default 0)."),
  }),
  z.object({
    op: z.literal("addCustomEvent"),
    id: z.string().optional().describe("Local handle for referencing this new node in later ops."),
    eventName: z.string().describe("Custom event name."),
    x: z.number().optional(),
    y: z.number().optional(),
  }),
  z.object({
    op: z.literal("connect"),
    from: z.string().describe("Source node: a local id from an earlier add op, or an existing node GUID."),
    fromPin: z.string().describe("Output pin name. Exec output is \"then\"."),
    to: z.string().describe("Target node: a local id or existing node GUID."),
    toPin: z.string().describe("Input pin name. Exec input is \"execute\"."),
  }),
  z.object({
    op: z.literal("setPinDefault"),
    node: z.string().describe("Node local id or GUID."),
    pin: z.string().describe("Input pin name (the UE parameter name, e.g. \"InString\")."),
    value: z.string().describe("Literal default value as a string."),
  }),
  z.object({
    op: z.literal("deleteNode"),
    node: z.string().describe("Node local id or GUID to delete (breaks its links)."),
  }),
  z.object({
    op: z.literal("addMemberVariable"),
    name: z.string().describe("Variable name to create on the blueprint."),
    varType: z
      .enum(["int", "bool", "float", "string", "name", "byte"])
      .default("int")
      .describe("Variable type."),
    default: z.string().optional().describe("Default value as a string, e.g. \"100\" or \"true\"."),
  }),
  z.object({
    op: z.literal("addVariableGet"),
    id: z.string().optional().describe("Local handle for referencing this node in later ops."),
    variable: z
      .string()
      .describe("Member/self variable name. The node's output data pin is named after the variable."),
    x: z.number().optional(),
    y: z.number().optional(),
  }),
  z.object({
    op: z.literal("addVariableSet"),
    id: z.string().optional().describe("Local handle for referencing this node in later ops."),
    variable: z
      .string()
      .describe("Member/self variable name. Exec in \"execute\"/out \"then\"; the input data pin (value to set) is named after the variable."),
    x: z.number().optional(),
    y: z.number().optional(),
  }),
  z.object({
    op: z.literal("addBranch"),
    id: z.string().optional().describe("Local handle for referencing this node in later ops."),
    x: z.number().optional(),
    y: z.number().optional(),
  }),
  z.object({
    op: z.literal("moveNode"),
    node: z.string().describe("Node local id or GUID to reposition."),
    x: z.number().describe("Graph X position."),
    y: z.number().describe("Graph Y position."),
  }),
  z.object({
    op: z.literal("addSequence"),
    id: z.string().optional().describe("Local handle for referencing this node in later ops."),
    numOutputs: z.number().int().min(2).default(2).describe("Number of exec output pins (then_0, then_1, ...)."),
    x: z.number().optional(),
    y: z.number().optional(),
  }),
  z.object({
    op: z.literal("addMacro"),
    id: z.string().optional().describe("Local handle for referencing this node in later ops."),
    macro: z.string().describe("Standard macro name: ForLoop, ForEachLoop, WhileLoop, Gate, DoOnce, FlipFlop."),
    x: z.number().optional(),
    y: z.number().optional(),
  }),
  z.object({
    op: z.literal("addCast"),
    id: z.string().optional().describe("Local handle for referencing this node in later ops."),
    targetClass: z
      .string()
      .describe("Class to cast to: a UE Python class name (e.g. \"PawnMovementComponent\") or a class path (\"/Game/.../BP_Foo.BP_Foo_C\")."),
    pure: z.boolean().default(false).describe("Pure cast (no exec pins)."),
    x: z.number().optional(),
    y: z.number().optional(),
  }),
  z.object({
    op: z.literal("addSelfFunctionNode"),
    id: z.string().optional().describe("Local handle for referencing this node in later ops."),
    functionName: z.string().describe("A function on this blueprint itself or an inherited/parent class."),
    x: z.number().optional(),
    y: z.number().optional(),
  }),
  z.object({
    op: z.literal("breakPinLink"),
    from: z.string().describe("Source node (local id or existing GUID)."),
    fromPin: z.string().describe("Output pin name."),
    to: z.string().describe("Target node (local id or existing GUID)."),
    toPin: z.string().describe("Input pin name."),
  }),
  z.object({
    op: z.literal("retargetNode"),
    node: z.string().describe("Existing node (local id or GUID) to reconfigure in place."),
    functionName: z.string().optional().describe("Retarget a CallFunction node to this function (pair with functionOwner)."),
    functionOwner: z.string().optional().describe("Owner class for functionName (UE Python class name or class path)."),
    targetClass: z.string().optional().describe("Retarget a Cast node to this class."),
    variable: z.string().optional().describe("Retarget a Variable Get/Set node to this self/member variable."),
    eventName: z.string().optional().describe("Rename a Custom Event node."),
  }),
]);

export function registerEditBlueprintGraphTool(server: McpServer): void {
  server.registerTool(
    "edit-blueprint-graph",
    {
      title: "Edit Blueprint Graph",
      description:
        "Add / wire / delete nodes in a Blueprint's K2 graph, then compile and save — all in one editor session. " +
        "Op kinds: addFunctionNode, addSelfFunctionNode, addCustomEvent, addVariableGet, addVariableSet, addBranch, addSequence, addMacro (ForLoop/ForEachLoop/WhileLoop/Gate/DoOnce/FlipFlop), addCast, addMemberVariable, connect, breakPinLink (disconnect one wire), setPinDefault, deleteNode, moveNode, retargetNode (reconfigure an existing node in place: retarget a function call / cast / variable, or rename a custom event). " +
        "Operations are applied in order; nodes spawned earlier in the batch can be referenced by a local `id` (later ops use that id or an existing node GUID). " +
        "Create a variable with addMemberVariable BEFORE adding get/set nodes that reference it. " +
        "Pass autoLayout: true to tidy the graph (or call with empty operations + autoLayout to just re-arrange an existing graph). " +
        "Requires the ClaudeUnrealBridge plugin enabled + its editor target built in the project. " +
        "Use read-blueprint first to get existing node/pin GUIDs. Destructive: modifies the .uasset on disk (use git to roll back).",
      inputSchema: {
        projectPath: z.string().describe("Absolute path to the UE project directory"),
        blueprintPath: z
          .string()
          .describe("Asset path (e.g. /Game/Blueprints/BP_Player) or file path to the blueprint"),
        graphName: z
          .string()
          .default("")
          .describe("Graph to edit by name. Empty (default) targets the event graph."),
        operations: z
          .array(opSchema)
          .default([])
          .describe("Ordered list of graph operations to apply in a single session. May be empty for a layout-only call (pair with autoLayout: true to re-arrange an existing graph)."),
        compile: z
          .boolean()
          .default(true)
          .describe("Compile the blueprint after applying operations (default true)."),
        autoLayout: z
          .boolean()
          .default(false)
          .describe(
            "Auto-arrange the whole graph into a tidy left-to-right flow after applying ops. Recommended when authoring a new graph; avoid on graphs a human has hand-arranged (it repositions every node).",
          ),
      },
      outputSchema: graphEditStructuredShape,
      annotations: {
        readOnlyHint: false,
        destructiveHint: true,
        idempotentHint: false,
        openWorldHint: false,
      },
    },
    async ({ projectPath, blueprintPath, graphName, operations, compile, autoLayout }, extra) => {
      try {
        if (operations.length === 0 && !autoLayout) {
          return {
            isError: true,
            content: [
              {
                type: "text" as const,
                text: "Nothing to do: `operations` is empty and `autoLayout` is false. Provide operations, or set autoLayout: true for a layout-only pass.",
              },
            ],
          };
        }
        const project = detectProject(projectPath);
        const assetPath = normalizeToAssetPath(blueprintPath, project);
        const scriptPath = join(pluginRoot, "python-scripts", "edit_blueprint_graph.py");

        const result = await runPythonInUE(
          project,
          scriptPath,
          {
            asset_path: assetPath,
            graph_name: graphName ?? "",
            operations: JSON.stringify(operations),
            compile: String(compile),
            auto_layout: String(autoLayout),
          },
          undefined,
          { signal: extra.signal, onProgress: progressFromExtra(extra) },
        );

        if (extra.signal?.aborted) {
          return { isError: true, content: [{ type: "text" as const, text: "Graph edit cancelled." }] };
        }

        if (!result.success || !result.data) {
          return {
            isError: true,
            content: [
              {
                type: "text" as const,
                text:
                  `Failed to edit blueprint graph.\n\nError: ${result.errorSummary ?? result.stderr ?? "Unknown error"}\n\n` +
                  `Common causes: the ClaudeUnrealBridge plugin isn't enabled/built in this project, the asset path is wrong, or the editor failed to load.`,
              },
            ],
          };
        }

        const data = result.data as GraphEditResult;

        // A library/asset-level failure surfaced as a top-level error string.
        if (data.error) {
          return {
            isError: true,
            content: [{ type: "text" as const, text: `Graph edit failed: ${data.error}` }],
          };
        }

        const structured = {
          success: data.success,
          assetPath: data.asset_path,
          graph: data.graph,
          compiled: data.compiled ?? null,
          saved: data.saved ?? null,
          autoLayout: data.auto_layout ?? null,
          handles: data.handles ?? {},
          operations: (data.operations ?? []).map((o) => ({
            index: o.index,
            op: o.op,
            ...(o.ok !== undefined ? { ok: o.ok } : {}),
            ...(o.guid ? { guid: o.guid } : {}),
            ...(o.error ? { error: o.error } : {}),
          })),
          ...(data.nodes ? { nodes: data.nodes } : {}),
          connections: (data.connections ?? [])
            .map(parseConnection)
            .filter((c): c is NonNullable<typeof c> => c !== null),
        };

        return {
          content: [{ type: "text" as const, text: formatResult(data) }],
          structuredContent: structured,
        };
      } catch (err) {
        return {
          isError: true,
          content: [
            {
              type: "text" as const,
              text: `Error: ${err instanceof Error ? err.message : String(err)}`,
            },
          ],
        };
      }
    },
  );
}

function formatResult(data: GraphEditResult): string {
  const lines: string[] = [];
  const head = data.success ? "OK" : "FAILED";
  lines.push(`## Edit graph: ${head} — \`${data.asset_path}\`${data.graph ? ` (graph: ${data.graph})` : " (event graph)"}`);

  lines.push("", "### Operations");
  for (const o of data.operations ?? []) {
    const tag = o.ok ? "✓" : "✗";
    const extra = o.guid ? ` → \`${o.guid}\`` : o.error ? ` — _${o.error}_` : "";
    lines.push(`${tag} ${o.index}. ${o.op}${extra}`);
  }

  const handleEntries = Object.entries(data.handles ?? {});
  if (handleEntries.length > 0) {
    lines.push("", "### Node handles");
    for (const [id, guid] of handleEntries) {
      lines.push(`- \`${id}\` = \`${guid}\``);
    }
  }

  lines.push("");
  if (data.compiled !== undefined && data.compiled !== null) {
    lines.push(`Compiled: ${data.compiled ? "✓" : "✗"}${data.compile_error ? ` (${data.compile_error})` : ""}`);
  }
  if (data.saved !== undefined && data.saved !== null) {
    lines.push(`Saved: ${data.saved ? "✓" : "✗"}${data.save_error ? ` (${data.save_error})` : ""}`);
  }
  if (data.auto_layout !== undefined && data.auto_layout !== null) {
    lines.push(`Auto-layout: ${data.auto_layout ? "✓" : "✗"}${data.auto_layout_error ? ` (${data.auto_layout_error})` : ""}`);
  }
  if (data.nodes) {
    lines.push(`Nodes now in graph: ${data.nodes.length}`);
  }
  if (data.connections) {
    lines.push(`Connections now in graph: ${data.connections.length}`);
  }

  return lines.join("\n");
}
